editor/utils/heartbeat.js

```diff
@@ -0,0 +1,165 @@
+import store from '../store';
+import {
+	getEditedPostTitle,
+	getEditedPostExcerpt,
+	getEditedPostContent,
+	getCurrentPostId,
+	isPostAutosaveDirty,
+} from '../store/selectors';
+import {
+	toggleAutosave,
+	removeNotice,
+	toggleNetworkIsConnected,
+	showDisconnectionNotice,
+	resetAutosave,
+	updateAutosave,
+	updateAutosaveStatusMessage,
+} from '../store/actions';
+
+import { compact } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { doAction } from '@wordpress/hooks';
+
+const DISCONNECTION_NOTICE_ID = 'DISCONNECTION_NOTICE_ID';
+
+/**
+ * Set up the heartbeat functionality for Gutenberg.
+ */
+export function setupHeartbeat() {
+	const $document = jQuery( document );
+
+	/**
+	 * Configure heartbeat to refresh the wp-api nonce, keeping the editor authorization intact.
+	 *
+	 * @todo update the _wpnonce used by autosaves.
+	 */
+	$document.on( 'heartbeat-tick', ( event, response ) => {
+		if ( response[ 'rest-nonce' ] ) {
+			window.wpApiSettings.nonce = response[ 'rest-nonce' ];
+		}
+	} );
+
+	/**
+	 * Configure Heartbeat autosaves.
+	 */
+	const { dispatch, getState } = store;
+
+	let compareString;
+	let lastCompareString;
+
+	/**
+	 * Autosave 'save' function that pulls content from Gutenberg state. Based on `wp.autosave.save`.
+	 *
+	 * @since 1.9.0
+	 *
+	 * @returns {Object|bool} postData The autosaved post data to send, or false if no autosave is needed.
+	 */
+	const save = function() {
+		// Bail early if autosaving is suspended or saving is blocked.
+		if ( wp.autosave.isSuspended || wp.autosave._blockSave ) {
+			return false;
+		}
+
+		// Check if its time for another autosave.
+		if ( ( new Date() ).getTime() < wp.autosave.nextRun ) {
+			return false;
+		}
+
+		// Get the current editor state and compute the compare string (title::excerpt::content).
+		const state = getState();
+
+		const toSend = {
+			post_title: getEditedPostTitle( state ),
+			post_excerpt: getEditedPostExcerpt( state ),
+			content: getEditedPostContent( state ),
+			post_id: getCurrentPostId( state ),
+		};
+
+		// Store the current editor values into the state autosave.
+		dispatch( resetAutosave( toSend ) );
```

Comment:
@youknowriad here when heartbeat fires, we store the current editor values into the autosave. if the values have changed since the last tick, state.autosave.isDirty will now be true, if nothing changed it will be false

Reply:
interestingly that wasn't the case in my testing, the dirty detection worked as expected. I may need some help here implementing this properly.

Reply:
@aduth can you offer a suggestion on how to refactor? I retested and it IS working as I expected it to.

I tested by logging `isPostAutosaveDirty( state ) )` before and after making some changes to the post and triggering a heartbeat:

![](https://cl.ly/1y2V1p0s011Z/Image%202018-02-20%20at%204.52.17%20PM.png)